pypesto/optimize/util.py

```diff
@@ -0,0 +1,70 @@
+from ..engine import Engine, SingleCoreEngine
+from ..objective import HistoryOptions
+from ..store.save_to_hdf5 import get_or_create_group
+from pathlib import Path
+
+
+import h5py
+
+
+def check_hdf5_mp(history_options: HistoryOptions,
+                  engine: Engine):
+    """
+    Create a folder for partial HDF5 files,
+    if a parallelization engine will be used.
+
+    Paramters
+    ---------
+    engine:
+        The Engine which is used in the optimization
+    history_options:
+        The HistoryOptions used in the optimization
+
+    Returns
+    -------
+    The filename that will be used to combine the partial HDF5 files later.
+    If a parallelization engine is not used, `None` is returned.
+    """
+    if isinstance(engine, SingleCoreEngine):
+        return None
+    filename = history_options.storage_file
+    file_path = Path(filename)
+
+    # create directory with same name as original file stem
+    partial_file_path = (
+            file_path.parent / file_path.stem /
+            (file_path.stem + '_{id}' + file_path.suffix)
+    )
+    partial_file_path.parent.mkdir(parents=True, exist_ok=True)
+    history_options.storage_file = str(partial_file_path)
+
+    # create hdf5 file that gathers the others within history group
+    with h5py.File(filename, mode='a') as f:
+        get_or_create_group(f, "history")
+    return filename
+
+
+def fill_hdf5_file(ret: list,
+                   filename: str):
+    """
+    Create links in `filename` to the
+    history of each start contained in ret, the results
+    of the optimization.
+
+    Parameters
+    ----------
+    ret:
+        The result iterable returned by the optimization.
+    filename:
+        The filename of the hdf5 file in which the histories
+        are to be gathered.
+    """
+    with h5py.File(filename, mode='a') as f:
+        for result in ret:
+            id = result['id']
+            if f'history/{id}' in f:
+                del f[f'history/{id}']
```

Comment:
this may be dangerous I think ... should this ever happen?

Reply:
mhh perhaps i should change that to a warning/error? Or could go with an additional `overwrite` argument?

Reply:
I would say error, no overwrite. there shouldn't be a scenario where one wants to overwrite previous results, right? when starting a new optimization, one should always use a new file, or alternatively append to an existing one, which must however be governed via the ids in the optimization already.

Reply:
What about doing an `overwrite ` flag as input and setting it to `False` as default...? Then it would be possible, but you would get an error, if the result already exists...

Reply:
I think that would make sense. Also a function (in the public API) to merge all history files into the final one would be useful, as this allows to just export a single file in the end.

Reply:
The `overwrite` flag was exactly what I meant 👍 will do it.
Would suggest adding the function  to merge files in a separate issue? would perhaps be nice either way if you want to merge two different optimization runs/results?

Reply:
was about to implement the overwrite function. But the `fill_hdf5_file` function is called in minimize. As this is basically only called if we use an hdf5 history with multiprocessengine, I do not think that this justifies the additional argument overwrite in minimize. Would then just remove the deletion part as the scenario, as yannik pointed out, should not really occur.